### ISSUE TITLE
Recell, Precision,F1 Score and Accuracy Saturation 

### ISSUE THREAD

**Issue by khawar512** [serengil/deepface#160]
I am running your library on 100 images again and again and every time the all values are saturated again and again. Do you have any reason why this phenomena happened ?

Precision | Recall | F1 Score | Accuracy
64.27 | 48.98 | 55.59 | 61.04
61.26 | 49.5 | 54.76 | 59.28
61.82 | 51.71 | 56.31 | 60.06

**Comment by serengil** [comment]
You are passing arrays as inputs to verify function? Could you share your code block?

**Comment by khawar512** [comment]
I am simply pass the list of images which is stored in json file. Last time, i just changed below part and put list.
`cm = confusion_matrix(list(y_test), predictions_classes)
`

**Comment by serengil** [comment]
how did you find the predictions_classes? I need that code block.

**Comment by khawar512** [comment]
```
positives = []

for key, values in idendities.items():
    print(key)
    for i in range(0, len(values) - 1):
        for j in range(i + 1, len(values)):
            # print(values[i], " and ", values[j])
            positive = [values[i], values[j]]
            positives.append(positive)

positives = pd.DataFrame(positives, columns=["file_x", "file_y"])
positives["decision"] = "Yes"
print(positives.shape)
# --------------------------
# Negatives

samples_list = list(idendities.values())

negatives = []

for i in range(0, len(idendities) - 1):
    for j in range(i + 1, len(idendities)):
        # print(samples_list[i], " vs ",samples_list[j])
        cross_product = itertools.product(samples_list[i], samples_list[j])
        cross_product = list(cross_product)
        # print(cross_product)

        for cross_sample in cross_product:
            # print(cross_sample[0], " vs ", cross_sample[1])
            negative = []
            negative.append(cross_sample[0])
            negative.append(cross_sample[1])
            negatives.append(negative)

negatives = pd.DataFrame(negatives, columns=["file_x", "file_y"])
negatives["decision"] = "No"

negatives = negatives.sample(positives.shape[0])

print(negatives.shape)
# --------------------------
# Merge positive and negative ones

df = pd.concat([positives, negatives]).reset_index(drop=True)

print(df.decision.value_counts())

df.file_x = "deepface/tests/dataset/" + df.file_x
df.file_y = "deepface/tests/dataset/" + df.file_y
# --------------------------
# DeepFace

from deepface import DeepFace
from deepface.basemodels import VGGFace, OpenFace, Facenet, FbDeepFace

pretrained_models = {}

pretrained_models["VGG-Face"] = VGGFace.loadModel()
print("VGG-Face loaded")
pretrained_models["Facenet"] = Facenet.loadModel()
print("Facenet loaded")
pretrained_models["OpenFace"] = OpenFace.loadModel()
print("OpenFace loaded")
pretrained_models["DeepFace"] = FbDeepFace.loadModel()
print("FbDeepFace loaded")

instances = df[["file_x", "file_y"]].values.tolist()

models = ['VGG-Face']
metrics = ['cosine']

if True:
    for model in models:
        for metric in metrics:

            resp_obj = DeepFace.verify(instances
                                       , model_name=model
                                       , model=pretrained_models[model]
                                       , distance_metric=metric)

            distances = []

            for i in range(0, len(instances)):
                distance = round(resp_obj["pair_%s" % (i + 1)]["distance"], 4)
                distances.append(distance)

            df['%s_%s' % (model, metric)] = distances

    df.to_csv("face-recognition-pivot.csv", index=False)
else:
    df = pd.read_csv("face-recognition-pivot.csv")

df_raw = df.copy()

# --------------------------
# Distribution

#fig = plt.figure(figsize=(15, 15))

#figure_idx = 1
#for model in models:
#    for metric in metrics:
#        feature = '%s_%s' % (model, metric)

#        ax1 = fig.add_subplot(4, 2, figure_idx)

#        df[df.decision == "Yes"][feature].plot(kind='kde', title=feature, label='Yes', legend=True)
#        df[df.decision == "No"][feature].plot(kind='kde', title=feature, label='No', legend=True)

#        figure_idx = figure_idx + 1

# plt.show()
# --------------------------
# Pre-processing for modelling

columns = []
for model in models:
    for metric in metrics:
        feature = '%s_%s' % (model, metric)
        columns.append(feature)

columns.append("decision")

df = df[columns]

df.loc[df[df.decision == 'Yes'].index, 'decision'] = 1
df.loc[df[df.decision == 'No'].index, 'decision'] = 0

print(df.head())
# --------------------------
# Train test split

from sklearn.model_selection import train_test_split

df_train, df_test = train_test_split(df, test_size=0.30, random_state=17)

target_name = "decision"

y_train = df_train[target_name].values
x_train = df_train.drop(columns=[target_name]).values

y_test = df_test[target_name].values
x_test = df_test.drop(columns=[target_name]).values

# --------------------------
# LightGBM

import lightgbm as lgb

features = df.drop(columns=[target_name]).columns.tolist()
lgb_train = lgb.Dataset(x_train, y_train, feature_name=features)
lgb_test = lgb.Dataset(x_test, y_test, feature_name=features)

params = {
    'task': 'train'
    , 'boosting_type': 'gbdt'
    , 'objective': 'multiclass'
    , 'num_class': 2
    , 'metric': 'multi_logloss'
}

gbm = lgb.train(params, lgb_train, num_boost_round=250, early_stopping_rounds=15, valid_sets=lgb_test)

gbm.save_model("face-recognition-ensemble-model.txt")

# --------------------------
# Evaluation

predictions = gbm.predict(x_test)

predictions_classes = []
for i in predictions:
    prediction_class = np.argmax(i)
    predictions_classes.append(prediction_class)

cm = confusion_matrix(list(y_test), predictions_classes)

tn, fp, fn, tp = cm.ravel()

recall = tp / (tp + fn)
precision = tp / (tp + fp)
accuracy = (tp + tn) / (tn + fp + fn + tp)
f1 = 2 * (precision * recall) / (precision + recall)

print("Precision: ", 100 * precision, "%")
print("Recall: ", 100 * recall, "%")
print("F1 score ", 100 * f1, "%")
print("Accuracy: ", 100 * accuracy, "%")
```

**Comment by serengil** [comment]
You build a model for just a feature (vggface and cosine pair). Why would you do that?

**Comment by khawar512** [comment]
I just want to deeply analyze one by one models like vggface and cosine pair and then another one with same format.

**Comment by serengil** [comment]
one feature is very weak for a gbm algorithm. I think that's the reason.

**Comment by khawar512** [comment]
But gbm is used in ensemble learning when we have two models. Am i right? Now we have only one model so i dont think so we need GBM. Am i right?

**Comment by serengil** [comment]
in ensemble, we use 4 models: vgg, facenet, openface, deepface; and their 3 metrics: euclidean, cosine, euclidean_l2. So there are 12 features in ensemble. if you just feed one model and one metric, then it would be one feature, and it is not meaningful for gbm.
